pkg/queue/forwarded_shim.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"net/http"
+	"strings"
+)
+
+// ForwardedShimHandler attempts to shim a `forwarded` http header from the information
+// available in the `x-forwarded-*` headers. When available, each node in the `x-forwarded-for`
+// header is combined with the `x-forwarded-proto` and `x-forwarded-host` fields to construct
+// a `forwarded` header. The `x-forwarded-by` header is ignored entirely, since it cannot be
+// reliably combined with `x-forwarded-for`. No-op if a `forwarded` header is already present.
+// Note: IPv6 addresses are *not* supported.
```

Comment:
I'm curious why? cc @tcnghia

Reply:
Assuming you're referring to the lack of IPv6 support, just laziness on my part. 

IPv4 addresses can be copied wholesale from `x-forwarded-for`, but IPv6 addresses need some extra processing (see https://tools.ietf.org/html/rfc7239#section-6).